patroni/api.py

```diff
@@ -908,21 +1493,48 @@ def __initialize(self, listen, ssl_options):
         if reloading_config:
             self.start()
 
-    def process_request_thread(self, request, client_address):
+    def process_request_thread(self, request: Union[socket.socket, SSLSocket], client_address: Tuple[str, int]) -> None:
+        """Process a request to the REST API.
+
+        Wrapper for :func:`ThreadingMixIn.process_request_thread` that additionally:
+            * Enable TCP keepalivel
+            * Perform SSL handshake (if an SSL socket).
+
+        :param request: socket to handle the client request.
+        :param client_address: tuple containing the client IP and port.
+        """
         enable_keepalive(request, 10, 3)
         if hasattr(request, 'context'):  # SSLSocket
-            request.do_handshake()
+            # pyright
+            if isinstance(request, SSLSocket):
+                request.do_handshake()
         super(RestApiServer, self).process_request_thread(request, client_address)
 
-    def shutdown_request(self, request):
+    def shutdown_request(self, request: Union[socket.socket, SSLSocket]) -> None:
+        """Shut down a request to the REST API.
+
+        Wrapper for :func:`HTTPServer.shutdown_request` that additionally:
+            * Perform SSL shutdown handshake (if a SSL socket).
+
+        :param request: socket to handle the client request.
+        """
         if hasattr(request, 'context'):  # SSLSocket
             try:
-                request.unwrap()
+                # pyright
+                if isinstance(request, SSLSocket):
+                    request.unwrap()
             except Exception as e:
                 logger.debug('Failed to shutdown SSL connection: %r', e)
         super(RestApiServer, self).shutdown_request(request)
 
-    def get_certificate_serial_number(self):
+    def get_certificate_serial_number(self) -> Union[str, None]:
+        """Get serial number of the certificate used by the REST API.
+
+        :returns: serial number of the certificate configured through ``restapi.certfile`` setting.
+        """
+        # pyright -- ``__ssl_options`` is initially created as ``None``, but right after that it is replaced with a
+        # dictionary through :func:`reload_config`.
+        assert isinstance(self.__ssl_options, dict)
         if self.__ssl_options.get('certfile'):
             import ssl
             try:
```

Comment:
imo this should be replaced with something that doesn't use the underlying `_ssl` shared library module (a protected member). e.g.

```
>>> from cryptography import x509
>>> c = x509.load_pem_x509_certificate(open(cert, 'rb').read())
>>> c.serial_number
4835703278459707669005204
```
I guess that depends on whether that cert is in pem format and it's a server cert. If this is used for client cert or der format then detection might be needed.

Reply:
There is a reason why it is done this way: https://github.com/zalando/patroni/pull/1887#discussion_r597204638

Reply:
Perhaps that reason could be reviewed?

```
$ grep crypto requirements.txt 
cryptography>=1.4
```

Reply:
Actually, it could be solved differently:
```python
ctx = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
crts = ctx.load_verify_locations(self.__ssl_options['certfile'])
if crts:
    return crts[0].get('serialNumber')

```